### PR TITLE
[Kinda WIP] Refactor groups for less crazy construction and timing updates

### PR DESCRIPTION
**Pull request by rjwright**
I have a feeling that this isn't 100% ready yet, but I've been working on it on-and-off for a while, so some outside perspective would be useful.

@shans PTAL

### CODE REVIEW

File: src/web-animations-next-player.js [hunk 1]
```diff
@@ -30,6 +30,38 @@
 
   // TODO: add a source getter/setter
   scope.Player.prototype = {
+    _updateChildTiming: function(childPlayer, offset) {
+      if (childPlayer.startTime != this.startTime + offset) {
+        if (this.startTime === null) {
```
Comment (shans) [inline src/web-animations-next-player.js]: Maybe while we're moving this we could look at swapping the order of these tests - if this.startTime == null, then this.startTime + offset = offset, which could potentially equal childPlayer.startTime, even if childPlayer.currentTime != this.source.player.currentTime - offset.

File: src/web-animations-next-player.js [hunk 2]
```diff
@@ -30,6 +30,38 @@
...
+          childPlayer.currentTime = this.source.player.currentTime - offset;
+          childPlayer._startTime = null;
+        } else {
+          childPlayer.startTime = this.startTime + offset;
+        }
+      }
+      // FIXME: What is this for?
```
Comment (shans) [inline src/web-animations-next-player.js]: Gooood question. What happens if we remove it?

**Comment by shans** [comment]
lg so far. Want to take a look @dstockwell?

File: src/web-animations-next-player.js (hunk 1 above)
Comment (rjwright) [inline src/web-animations-next-player.js]: > Maybe while we're moving this we could look at swapping the order of these tests - if this.startTime == null, then this.startTime + offset = offset, which could potentially equal childPlayer.startTime, even if childPlayer.currentTime != this.source.player.currentTime - offset.

Will look into it. I will also double-check to see if it's reasonable to remove all Player.proto members from animation-constructor.js.

Comment (rjwright) [inline src/web-animations-next-player.js]: > Maybe while we're moving this we could look at swapping the order of these tests - if this.startTime == null, then this.startTime + offset = offset, which could potentially equal childPlayer.startTime, even if childPlayer.currentTime != this.source.player.currentTime - offset.

Wait, we don't want to alter this.startTime here at all. All _updateChildTiming is doing now is scheduling children relative to the parent (taking into account any delay on the parent, plus arranging children within a sequence), and doing the thing that we don't know the purpose of.

If we can delete the mysterious code I will rename this method to something like "arrangeChildren".

Comment (rjwright) [inline src/web-animations-next-player.js]: > If we can delete the mysterious code I will rename this method to something like "arrangeChildren".

Actually I think I'll do this anyway

File: src/web-animations-next-player.js (hunk 2 above)
Comment (rjwright) [inline src/web-animations-next-player.js]: > Gooood question. What happens if we remove it?

Nothing, but I don't know if it is covered by unit tests.

I tested just with a little groups demo, seeking it to the end and then reversing it. That *looks* fine. I don't know if its state is correct w.r.t. to the model throughout.

It seems that if this code is useful for something it should be if (this.playbackRate < 0) not if (this.playbackRate == -1).

I'm guessing what it's meant to do is set the child player currentTime to -1 when the group is playing in reverse and the child is "done" (past its beginning). I don't know why we would want to do that. I also think that if it's a thing that needs doing it should happen in forEachChild, or a ticker, or both.

Comment (shans) [inline src/web-animations-next-player.js]: OK, can you add a test when the group is playing in reverse with 'none' fill, and the child plays past its beginning (i.e. unfills)? If that passes without this bit of code, nuke the code.

File: src/web-animations-next-player.js [hunk 3]
```diff
@@ -47,6 +47,63 @@
...
+        child.player = this.source.player;
+
+        this._arrangeChildren(childPlayer, offset);
+
+        if (this.source instanceof window.AnimationSequence)
+          offset += scope.groupChildDuration(child);
+      }.bind(this));
+    },
+    _arrangeChildren: function(childPlayer, offset) {
+      if (childPlayer.startTime != this.startTime + offset) {
```
Comment (shans) [inline src/web-animations-next-player.js]: This still has the same problem. If this.startTime is null, then this.startTime  + offset = offset, which could accidentally be the same as childPlayer.startTime. But if this.startTime is null then we want to set childPlayer.startTime to null.

File: src/web-animations-next-player.js [hunk 4]
```diff
@@ -47,6 +47,63 @@
...
+        if (this.startTime === null) {
```
Comment (dstockwell) [inline src/web-animations-next-player.js]: Did you mean to add null on the previous line? Maybe make this the first check.

File: src/web-animations-next-player.js [hunk 5]
```diff
@@ -157,11 +214,22 @@
...
       var offset = 0;
+      if (this.source.children && this._childPlayers.length < this.source.children.length)
+        this._constructChildren();
       this._childPlayers.forEach(function(child) {
         f.call(this, child, offset);
         if (this.source instanceof window.AnimationSequence)
           offset += child.source.activeDuration;
       }.bind(this));
+
+      if (this._player.playState == 'pending')
```
Comment (dstockwell) [inline src/web-animations-next-player.js]: This logic doesn't belong in `forEachChild`. Rename or move it to the caller.
